### ISSUE TITLE
Parallel rendering with glb files

### ISSUE THREAD

**Issue by JayeFu** [mosra/magnum#629]
Hi,

We have some *.glb files to render with magnum. Is there a functionality that we can use for parallel rendering (say, 20 different viewpoints in the same scene in parallel) in python?

Many thanks in advance!

Best,
Jiawei

**Comment by mosra** [comment]
Hi, a wild guess before we go into details, to avoid wheel reinvention -- is this in any way related to habitat-sim, or not at all? :) My answer will depend on that.

**Comment by JayeFu** [comment]
Thanks a lot for the reply!

The `gfx_batch` seems to be very attractive. We are looking into it now.

If we choose to implement the stuff with plain Magnum, could you please point us to some pointers about the whole pipeline of using Magnum Python API for rendering .glb files? I guess we could also try directly parallelizing that.

Many thanks in advance!

**Comment by mosra** [comment]
The Python API isn't complete enough for batch-renderer-like workflow yet, you'll need to use the C++ APIs directly. A high-level rendering process and asset constraints are hopefully explained clear enough in the docs I linked, and I tried my best to make the `gfx::batch::Renderer` internals documented and clear to follow as well.

According to my benchmarks, the scene setup & submission process in the batch renderer itself wasn't really a bottleneck where any parallelization would help much, most of the time was spent waiting on the GPU, and for example the transformation processing was just about 2% of the frame time. With the pre-processed assets there isn't really any other complexity happening that would need to be heavily parallelized to achieve substantial speed improvements. What was slow in comparison was the high-level (Python) code managing the scenes, physics, etc.

Have fun, and let's use the habitat-sim issue tracker for future questions re the batch renderer :)
